File: cses/1. Introductory Problems/Palindrome Reorder/solution.cc
#include <bits/stdc++.h>

using namespace std;

int main() {
    unordered_map<char, int> hash;
    int cnt = 0;
    char x, odd_val = ' ';
    while (cin >> x) {
        ++hash[x];
        ++cnt;
    }
    bool valid = true;
    int odd = 0;
    for (auto &[c, v] : hash) {
        if (v % 2) {
            ++odd;
            odd_val = c;
            if (cnt % 2 == 0 || odd >= 2) {
                valid = false;
                break;
            }
        } 
    }
    if (valid) {
        string res = "";
        for (auto &[c, v] : hash) {
            res += string(v / 2, c);
        }
        string temp(res.rbegin(), res.rend());
        if (odd_val != ' ') res += odd_val;
        res += temp;
        cout << res << endl;
    } else {
        cout << "NO SOLUTION" << endl;
    }
    return 0;
}

File: cses/1. Introductory Problems/Palindrome Reorder/solution.py
s = input()
n = len(s)
d = {}
for c in s:
    if c in d:
        d[c] += 1
    else:
        d[c] = 1

valid = True
odd_cnt = 0
odd_val = ''
for c, v in d.items():
    if v % 2 == 1:
        odd_cnt += 1
        odd_val = c
        if n % 2 == 0 or odd_cnt == 2:
            valid = False
            break
if not valid:
    print("NO SOLUTION")
else:
    res = ''
    for c, v in d.items():
        res += c * (v // 2)
    temp = res[::-1]
    res += odd_val
    res += temp
    print(res)
